cache.go

```diff
@@ -27,26 +29,37 @@ func addClientCache(clientIp string) {
 	clientSpecificCachesMutex.Unlock()
 }
 
-func globalCacheHit(req *dns.Msg) *dns.Msg {
+func globalCacheHit(req *dns.Msg) (*dns.Msg, time.Time) {
 	return globalCache.Hit(req.Question[0], false, false, req.MsgHdr.Id)
 }
 
-func clientSpecificCacheHit(clientIp string, req *dns.Msg) *dns.Msg {
+func clientSpecificCacheHit(clientIp string, req *dns.Msg) (*dns.Msg, time.Time) {
 	addClientCache(clientIp)
 	clientCache := getClientCache(clientIp)
 	return clientCache.Hit(req.Question[0], false, false, req.MsgHdr.Id)
 }
 
 func addToGlobalCache(req, msg *dns.Msg) {
 	key := cache.Key(req.Question[0], false, false)
-	globalCache.InsertMessage(key, msg)
+
+	var globalTtl = globalCache.GetTTL()
+	var requestTtl = globalTtl
+
+	if len(msg.Answer) > 0 {
+		requestTtl = time.Duration(msg.Answer[0].Header().Ttl) * time.Second
+	}
+	if requestTtl < globalTtl {
+		globalCache.InsertMessage(key, msg, requestTtl)
+	} else {
+		globalCache.InsertMessage(key, msg, globalTtl)
+	}
 }
 
 func addToClientSpecificCache(clientIp string, req, msg *dns.Msg) {
 	addClientCache(clientIp)
 	clientCache := getClientCache(clientIp)
 	key := cache.Key(req.Question[0], false, false)
-	clientCache.InsertMessage(key, msg)
+	clientCache.InsertMessage(key, msg, clientCache.GetTTL())
```

Comment:
This should have the same logic as addToGlobalCache for comparing request TTL vs cache.GetTTL.  Or they could really be combined into one `addToCache(cache, question, answer)` method that just takes in the cache to use.

A record could potentially say it has a TTL of 2^31 seconds, but we don't want to cache it for 68 years.